File: packages/deno/src/components.ts
import type { XmlElement } from "./xml.ts";
import type { Block, State } from "./types.ts";
import { evalExpr, interpolate } from "./expr.ts";
import { getStyleProp } from "./style.ts";
import { stringify as yamlStringify } from "@std/yaml";

// ---------------------------------------------------------------------------
// Document component
// ---------------------------------------------------------------------------

export async function processDocument(
  node: XmlElement,
  state: State,
): Promise<Block[]> {
  const rawSrc = node.attrs.src;
  if (!rawSrc || !state.filePath) return [];
  const src = interpolate(rawSrc, state.ctx);

  const dir = state.filePath.substring(0, state.filePath.lastIndexOf("/") + 1);
  const docPath = dir + src;
  let content: string;
  try {
    content = await Deno.readTextFile(docPath);
  } catch {
    return [];
  }

  return [{ type: "paragraph", text: content }];
}

// ---------------------------------------------------------------------------
// Table component
// ---------------------------------------------------------------------------

export async function processTable(
  node: XmlElement,
  state: State,
): Promise<Block[]> {
  const syntax = node.attrs.syntax ?? getStyleProp("table", "syntax", state) ?? "markdown";

  let records: Record<string, unknown>[] = [];
  let columns: string[] = [];

  // Check for inline records attribute first
  const recordsExpr = node.attrs.records;
  if (recordsExpr) {
    const raw = recordsExpr.replace(/^\{\{(.+)\}\}$/, "$1").trim();
    const data = evalExpr(raw, state.ctx);
    if (Array.isArray(data) && data.length > 0) {
      const first = data[0];
      if (Array.isArray(first)) {
        // Array of arrays — generate "Column N" headers
        records = data.map((row: unknown[]) => {
          const obj: Record<string, unknown> = {};
          row.forEach((val, i) => { obj[`Column ${i}`] = val; });
          return obj;
        });
        columns = Object.keys(records[0]);
      } else if (first !== null && typeof first === "object") {
        columns = Object.keys(first as Record<string, unknown>);
        records = data as Record<string, unknown>[];
      }
    }
  } else {
    // File-based source
    const rawSrc = node.attrs.src;
    if (!rawSrc || !state.filePath) return [];
    const src = interpolate(rawSrc, state.ctx);

    const dir = state.filePath.substring(0, state.filePath.lastIndexOf("/") + 1);
    const tablePath = dir + src;
    const ext = src.substring(src.lastIndexOf(".") + 1).toLowerCase();

    if (ext === "csv") {
      const content = await Deno.readTextFile(tablePath);
      const parsed = parseCsvTable(content);
      records = parsed.records;
      columns = parsed.columns;
    } else if (ext === "json") {
      const content = await Deno.readTextFile(tablePath);
      const data = JSON.parse(content);
      if (Array.isArray(data) && data.length > 0) {
        columns = Object.keys(data[0]);
        records = data;
      }
    } else if (ext === "jsonl" || (node.attrs.parser === "jsonl")) {
      const content = await Deno.readTextFile(tablePath);
      const parsed = parseJsonlTable(content);
      records = parsed.records;
      columns = parsed.columns;
    } else {
      return [];
    }

    // Apply type inference to CSV records
    records = inferColumnTypes(records, columns);
  }

  // Apply columns attribute (field/header definitions)
  const columnsAttr = node.attrs.columns;
  const headerMap: Record<string, string> = {};
  if (columnsAttr) {
    const raw = columnsAttr.replace(/^\{\{(.+)\}\}$/, "$1").trim();
    const colDefs = evalExpr(raw, state.ctx);
    if (Array.isArray(colDefs)) {
      const fields = colDefs.map((d: { field: string }) => d.field);
      columns = fields;
      for (const d of colDefs) {
        if (d.header) headerMap[d.field] = d.header;
      }
    }
  }

  // Apply selectedRecords slice
  const selectedRecords = node.attrs.selectedRecords;
  if (selectedRecords) {
    const [start, end] = parsePythonStyleSlice(selectedRecords, records.length);
    records = records.slice(start, end);
  }

  // Apply selectedColumns (only for file-based sources, not records attribute)
  const selectedColumns = node.attrs.selectedColumns;
  if (selectedColumns && !recordsExpr) {
    const cols = selectedColumns.split(",").map((c: string) => c.trim());
    columns = cols.filter((c: string) => columns.includes(c));
  }

  // Apply maxRecords — show first + "..." + last when truncating
  const maxRecords = parseInt(node.attrs.maxRecords ?? "");
  if (!isNaN(maxRecords) && maxRecords > 0 && records.length > maxRecords) {
    const first = records.slice(0, maxRecords - 1);
    const last = records[records.length - 1];
    const ellipsisRow: Record<string, unknown> = {};
    for (const c of columns) ellipsisRow[c] = "...";
    records = [...first, ellipsisRow, last];
  }

  if (columns.length === 0 || records.length === 0) return [];

  // Format cell: booleans → empty (React drops booleans); numbers → string
  const fmtCell = (v: unknown): string => {
    if (typeof v === "boolean") return "";
    if (v == null) return "";
    return String(v);
  };

  // Get writerOptions from node attrs (merged by applyStylesheet) or stylesheet
  // deno-lint-ignore no-explicit-any
  let writerOptions: Record<string, any> = {};
  // deno-lint-ignore no-explicit-any
  const rawWO = node.attrs.writerOptions ?? (state.styles?.["table"] as any)?.["writerOptions"];
  if (rawWO) {
    if (typeof rawWO === "string") {
      try { writerOptions = JSON.parse(rawWO); } catch { /* ignore */ }
    } else if (typeof rawWO === "object") {
      writerOptions = rawWO;
    }
  }

  if (syntax === "csv") {
    const separator = writerOptions.csvSeparator ?? ",";
    const showHeader = writerOptions.csvHeader !== false;
    const lines: string[] = [];
    if (showHeader) {
      lines.push(columns.join(separator));
    }
    for (const rec of records) {
      lines.push(columns.map((c) => fmtCell(rec[c])).join(separator));
    }
    return [{ type: "paragraph", text: lines.join("\n") }];
  }

  if (syntax === "tsv") {
    const headerLine = columns.join("\t");
    const bodyLines = records.map(
      (rec) => columns.map((c) => fmtCell(rec[c])).join("\t"),
    );
    const table = [headerLine, ...bodyLines].join("\n");
    return [{ type: "paragraph", text: table }];
  }

  // Default: markdown table
  const displayHeaders = columns.map((c) => headerMap[c] ?? c);

  const colWidths = columns.map((col, idx) => {
    let max = displayHeaders[idx].length;
    for (const rec of records) {
      const val = fmtCell(rec[col]);
      if (val.length > max) max = val.length;
    }
    return Math.max(max, 3);
  });

  const pad = (s: string, w: number) => s + " ".repeat(Math.max(0, w - s.length));
  const headerLine = "| " + displayHeaders.map((h, i) => pad(h, colWidths[i])).join(" | ") + " |";
  const sepLine = "| " + colWidths.map((w) => "-".repeat(w)).join(" | ") + " |";
  const bodyLines = records.map(
    (rec) => "| " + columns.map((c, i) => pad(fmtCell(rec[c]), colWidths[i])).join(" | ") + " |",
  );
  const table = [headerLine, sepLine, ...bodyLines].join("\n");
  return [{ type: "paragraph", text: table }];
}

// ---------------------------------------------------------------------------
// Object component — serializes data as XML or JSON in a code fence
// ---------------------------------------------------------------------------

export function processObject(
  node: XmlElement,
  state: State,
): Block[] {
  const dataExpr = node.attrs.data;
  if (!dataExpr) return [];
  const raw = dataExpr.replace(/^\{\{(.+)\}\}$/, "$1").trim();
  const data = evalExpr(raw, state.ctx);
  if (data == null) return [];

  const syntax = node.attrs.syntax ?? "json";

  if (syntax === "yaml") {
    const yaml = yamlStringify(data as Record<string, unknown>, { lineWidth: -1 }).trimEnd();
    return [{ type: "paragraph", text: "```yaml\n" + yaml + "\n```" }];
  }

  if (syntax === "xml") {
    const xml = objectToXml(data);
    return [{ type: "paragraph", text: "```xml\n" + xml + "\n```" }];
  }
  // Default: JSON
  const json = JSON.stringify(data, null, 2);
  return [{ type: "paragraph", text: "```json\n" + json + "\n```" }];
}

// ---------------------------------------------------------------------------
// Tree component — renders a tree structure as markdown headings
// ---------------------------------------------------------------------------

interface TreeItemData {
  name: string;
  children?: TreeItemData[];
  value?: string;
}

export function processTree(
  node: XmlElement,
  state: State,
): Block[] {
  const itemsExpr = node.attrs.items;
  if (!itemsExpr) return [];
  const raw = itemsExpr.replace(/^\{\{(.+)\}\}$/, "$1").trim();
  const items = evalExpr(raw, state.ctx) as TreeItemData[];
  if (!Array.isArray(items)) return [];

  const showContent = node.attrs.showContent === "true" || node.attrs.showContent as unknown === true;
  const blocks: Block[] = [];

  function walkTree(nodes: TreeItemData[], depth: number, pathPrefix: string): void {
    for (const item of nodes) {
      const fullPath = pathPrefix ? `${pathPrefix}/${item.name}` : item.name;
      blocks.push({ type: "heading", depth, text: fullPath });

      if (showContent && item.value !== undefined) {
        const ext = item.name.includes(".") ? item.name.substring(item.name.lastIndexOf(".") + 1) : "";
        const fence = "```" + ext + "\n" + item.value + "\n```";
        blocks.push({ type: "paragraph", text: fence });
      }

      if (item.children) {
        walkTree(item.children, depth + 1, fullPath);
      }
    }
  }

  walkTree(items, state.depth, "");
  return blocks;
}

// ---------------------------------------------------------------------------
// Shared helpers (also used by read.ts for <conversation>)
// ---------------------------------------------------------------------------

export function parsePythonStyleSlice(slice: string, totalLength: number): [number, number] {
  if (slice === ":") return [0, totalLength];
  if (slice.endsWith(":")) return [parseInt(slice.slice(0, -1)), totalLength];
  if (slice.startsWith(":")) {
    const end = parseInt(slice.slice(1));
    return [0, end < 0 ? totalLength + end : end];
  }
  if (slice.includes(":")) {
    const [s, e] = slice.split(":").map(Number);
    return [s, e < 0 ? totalLength + e : e];
  }
  const index = parseInt(slice);
  return [index, index + 1];
}

// ---------------------------------------------------------------------------
// Internal helpers
// ---------------------------------------------------------------------------

function parseCsvTable(content: string): { records: Record<string, string>[]; columns: string[] } {
  const lines = content.trim().split("\n");
  if (lines.length === 0) return { records: [], columns: [] };
  const headers = lines[0].split(",").map((h) => h.trim());
  const records = lines.slice(1).map((line) => {
    const vals = line.split(",").map((v) => v.trim());
    const obj: Record<string, string> = {};
    headers.forEach((h, i) => { obj[h] = vals[i] ?? ""; });
    return obj;
  });
  return { records, columns: headers };
}

function parseJsonlTable(content: string): { records: Record<string, unknown>[]; columns: string[] } {
  const lines = content.trim().split("\n").filter((l) => l.trim());
  const records = lines.map((line) => JSON.parse(line) as Record<string, unknown>);
  const columns = records.length > 0 ? Object.keys(records[0]) : [];
  return { records, columns };
}

function inferColumnTypes(
  records: Record<string, unknown>[],
  columns: string[],
): Record<string, unknown>[] {
  if (records.length === 0) return records;

  const colTypes: Record<string, string> = {};
  for (const col of columns) {
    let type = "string";
    let allEmpty = true;
    for (const rec of records) {
      const val = String(rec[col] ?? "").trim();
      if (val === "") continue;
      allEmpty = true;
      allEmpty = false;
      if (val === "true" || val === "false") {
        if (type === "string") type = "boolean";
      } else if (!isNaN(Number(val)) && val !== "") {
        if (type === "string") {
          type = parseFloat(val) === parseInt(val) ? "integer" : "float";
        }
      } else {
        type = "string";
        break;
      }
    }
    if (allEmpty) type = "string";
    colTypes[col] = type;
  }

  return records.map((rec) => {
    const out: Record<string, unknown> = {};
    for (const col of columns) {
      const raw = String(rec[col] ?? "").trim();
      switch (colTypes[col]) {
        case "boolean":
          out[col] = raw === "true";
          break;
        case "integer":
          out[col] = raw === "" ? "" : parseInt(raw);
          break;
        case "float":
          out[col] = raw === "" ? "" : parseFloat(raw);
          break;
        default:
          out[col] = raw;
      }
    }
    return out;
  });
}

/** Convert a JS value to XML string (mimics React-like serialization). */
function objectToXml(data: unknown, indent = "", wrapInItem = false): string {
  if (Array.isArray(data)) {
    return data.map((item) => objectToXml(item, indent, true)).join("\n");
  }
  if (data !== null && typeof data === "object") {
    const obj = data as Record<string, unknown>;
    const keys = Object.keys(obj);
    const childIndent = wrapInItem ? indent + "  " : indent;
    const inner = keys.map((k) => {
      const val = obj[k];
      if (val !== null && typeof val === "object") {
        return `${childIndent}<${k}>\n${objectToXml(val, childIndent + "  ")}\n${childIndent}</${k}>`;
      }
      const strVal = String(val ?? "");
      if (strVal === "") {
        return `${childIndent}<${k}/>`;
      }
      return `${childIndent}<${k}>${escapeXmlContent(strVal)}</${k}>`;
    }).join("\n");
    if (wrapInItem) {
      return `${indent}<item>\n${inner}\n${indent}</item>`;
    }
    return inner;
  }
  return indent + escapeXmlContent(String(data ?? ""));
}

function escapeXmlContent(s: string): string {
  return s.replace(/&/g, "&amp;").replace(/</g, "&lt;").replace(/>/g, "&gt;");
}

File: packages/python/pomlight/components.py
from __future__ import annotations

import json
from pathlib import Path
from typing import Any

import yaml

from .xml_parser import XmlElement
from .types import Block, Paragraph, State
from .expr import eval_expr, interpolate
from .style import get_style_prop


class _IndentedListDumper(yaml.SafeDumper):
    """Custom YAML dumper that indents list items inside mappings (matching Deno @std/yaml)."""
    def increase_indent(self, flow: bool = False, indentless: bool = False) -> None:
        return super().increase_indent(flow, False)


# ---------------------------------------------------------------------------
# Document component
# ---------------------------------------------------------------------------

def process_document(node: XmlElement, state: State) -> list[Block]:
    raw_src = node.attrs.get("src")
    if not raw_src or not state.file_path:
        return []
    src = interpolate(raw_src, state.ctx)

    dir_path = state.file_path[: state.file_path.rfind("/") + 1]
    doc_path = dir_path + src
    try:
        content = Path(doc_path).read_text()
    except OSError:
        return []

    return [Paragraph(text=content)]


# ---------------------------------------------------------------------------
# Table component
# ---------------------------------------------------------------------------

def process_table(node: XmlElement, state: State) -> list[Block]:
    syntax = node.attrs.get("syntax") or get_style_prop("table", "syntax", state) or "markdown"

    records: list[dict[str, Any]] = []
    columns: list[str] = []

    records_expr = node.attrs.get("records")
    if records_expr:
        raw = records_expr.strip()
        if raw.startswith("{{") and raw.endswith("}}"):
            raw = raw[2:-2].strip()
        data = eval_expr(raw, state.ctx)
        if isinstance(data, list) and len(data) > 0:
            first = data[0]
            if isinstance(first, list):
                records = [
                    {f"Column {i}": val for i, val in enumerate(row)}
                    for row in data
                ]
                columns = list(records[0].keys())
            elif isinstance(first, dict):
                columns = list(first.keys())
                records = data
    else:
        raw_src = node.attrs.get("src")
        if not raw_src or not state.file_path:
            return []
        src = interpolate(raw_src, state.ctx)

        dir_path = state.file_path[: state.file_path.rfind("/") + 1]
        table_path = dir_path + src
        ext = src.rsplit(".", 1)[-1].lower() if "." in src else ""

        if ext == "csv":
            try:
                content = Path(table_path).read_text()
            except OSError:
                return []
            parsed = _parse_csv_table(content)
            records = parsed["records"]
            columns = parsed["columns"]
        elif ext == "json":
            try:
                content = Path(table_path).read_text()
            except OSError:
                return []
            data = json.loads(content)
            if isinstance(data, list) and len(data) > 0:
                columns = list(data[0].keys())
                records = data
        elif ext == "jsonl" or node.attrs.get("parser") == "jsonl":
            try:
                content = Path(table_path).read_text()
            except OSError:
                return []
            parsed = _parse_jsonl_table(content)
            records = parsed["records"]
            columns = parsed["columns"]
        else:
            return []

        records = _infer_column_types(records, columns)

    # Apply columns attribute
    columns_attr = node.attrs.get("columns")
    header_map: dict[str, str] = {}
    if columns_attr:
        raw = columns_attr.strip()
        if raw.startswith("{{") and raw.endswith("}}"):
            raw = raw[2:-2].strip()
        col_defs = eval_expr(raw, state.ctx)
        if isinstance(col_defs, list):
            fields = [d["field"] for d in col_defs]
            columns = fields
            for d in col_defs:
                if d.get("header"):
                    header_map[d["field"]] = d["header"]

    # Apply selectedRecords slice
    selected_records = node.attrs.get("selectedRecords")
    if selected_records:
        start, end = parse_python_style_slice(selected_records, len(records))
        records = records[start:end]

    # Apply selectedColumns
    selected_columns = node.attrs.get("selectedColumns")
    if selected_columns and not records_expr:
        cols = [c.strip() for c in selected_columns.split(",")]
        columns = [c for c in cols if c in columns]

    # Apply maxRecords
    max_records_str = node.attrs.get("maxRecords", "")
    try:
        max_records = int(max_records_str)
    except (ValueError, TypeError):
        max_records = 0
    if max_records > 0 and len(records) > max_records:
        first = records[: max_records - 1]
        last = records[-1]
        ellipsis_row: dict[str, Any] = {c: "..." for c in columns}
        records = [*first, ellipsis_row, last]

    if not columns or not records:
        return []

    def fmt_cell(v: Any) -> str:
        if isinstance(v, bool):
            return ""
        if v is None:
            return ""
        return str(v)

    # Writer options
    writer_options: dict[str, Any] = {}
    raw_wo = node.attrs.get("writerOptions")
    if raw_wo is None and state.styles and "table" in state.styles:
        table_style = state.styles["table"]
        if isinstance(table_style, dict):
            raw_wo = table_style.get("writerOptions")
    if raw_wo:
        if isinstance(raw_wo, str):
            try:
                writer_options = json.loads(raw_wo)
            except (json.JSONDecodeError, TypeError):
                pass
        elif isinstance(raw_wo, dict):
            writer_options = raw_wo

    if syntax == "csv":
        separator = writer_options.get("csvSeparator", ",")
        show_header = writer_options.get("csvHeader", True) is not False
        lines: list[str] = []
        if show_header:
            lines.append(separator.join(columns))
        for rec in records:
            lines.append(separator.join(fmt_cell(rec.get(c)) for c in columns))
        return [Paragraph(text="\n".join(lines))]

    if syntax == "tsv":
        header_line = "\t".join(columns)
        body_lines = ["\t".join(fmt_cell(rec.get(c)) for c in columns) for rec in records]
        table = "\n".join([header_line, *body_lines])
        return [Paragraph(text=table)]

    # Default: markdown table
    display_headers = [header_map.get(c, c) for c in columns]

    col_widths: list[int] = []
    for idx, col in enumerate(columns):
        max_w = len(display_headers[idx])
        for rec in records:
            val = fmt_cell(rec.get(col))
            if len(val) > max_w:
                max_w = len(val)
        col_widths.append(max(max_w, 3))

    def pad(s: str, w: int) -> str:
        return s + " " * max(0, w - len(s))

    header_line = "| " + " | ".join(pad(h, col_widths[i]) for i, h in enumerate(display_headers)) + " |"
    sep_line = "| " + " | ".join("-" * w for w in col_widths) + " |"
    body_lines = [
        "| " + " | ".join(pad(fmt_cell(rec.get(c)), col_widths[i]) for i, c in enumerate(columns)) + " |"
        for rec in records
    ]
    table = "\n".join([header_line, sep_line, *body_lines])
    return [Paragraph(text=table)]


# ---------------------------------------------------------------------------
# Object component
# ---------------------------------------------------------------------------

def process_object(node: XmlElement, state: State) -> list[Block]:
    data_expr = node.attrs.get("data")
    if not data_expr:
        return []
    raw = data_expr.strip()
    if raw.startswith("{{") and raw.endswith("}}"):
        raw = raw[2:-2].strip()
    data = eval_expr(raw, state.ctx)
    if data is None:
        return []

    syntax = node.attrs.get("syntax", "json")

    if syntax == "yaml":
        yaml_str = yaml.dump(data, default_flow_style=False, width=1000000, sort_keys=False, Dumper=_IndentedListDumper).rstrip()
        return [Paragraph(text="```yaml\n" + yaml_str + "\n```")]

    if syntax == "xml":
        xml_str = _object_to_xml(data)
        return [Paragraph(text="```xml\n" + xml_str + "\n```")]

    # Default: JSON
    json_str = json.dumps(data, indent=2)
    return [Paragraph(text="```json\n" + json_str + "\n```")]


# ---------------------------------------------------------------------------
# Tree component
# ---------------------------------------------------------------------------

def process_tree(node: XmlElement, state: State) -> list[Block]:
    from .types import Heading

    items_expr = node.attrs.get("items")
    if not items_expr:
        return []
    raw = items_expr.strip()
    if raw.startswith("{{") and raw.endswith("}}"):
        raw = raw[2:-2].strip()
    items = eval_expr(raw, state.ctx)
    if not isinstance(items, list):
        return []

    show_content = node.attrs.get("showContent") in ("true", True)
    blocks: list[Block] = []

    def walk_tree(nodes: list[dict[str, Any]], depth: int, path_prefix: str) -> None:
        for item in nodes:
            name = item.get("name", "")
            full_path = f"{path_prefix}/{name}" if path_prefix else name
            blocks.append(Heading(depth=depth, text=full_path))

            if show_content and "value" in item:
                ext = name.rsplit(".", 1)[-1] if "." in name else ""
                fence = f"```{ext}\n{item['value']}\n```"
                blocks.append(Paragraph(text=fence))

            if "children" in item:
                walk_tree(item["children"], depth + 1, full_path)

    walk_tree(items, state.depth, "")
    return blocks


# ---------------------------------------------------------------------------
# Shared helpers
# ---------------------------------------------------------------------------

def parse_python_style_slice(slice_str: str, total_length: int) -> tuple[int, int]:
    if slice_str == ":":
        return (0, total_length)
    if slice_str.endswith(":"):
        return (int(slice_str[:-1]), total_length)
    if slice_str.startswith(":"):
        end = int(slice_str[1:])
        return (0, total_length + end if end < 0 else end)
    if ":" in slice_str:
        parts = slice_str.split(":")
        s, e = int(parts[0]), int(parts[1])
        return (s, total_length + e if e < 0 else e)
    index = int(slice_str)
    return (index, index + 1)


# ---------------------------------------------------------------------------
# Internal helpers
# ---------------------------------------------------------------------------

def _parse_csv_table(content: str) -> dict[str, Any]:
    lines = content.strip().split("\n")
    if not lines:
        return {"records": [], "columns": []}
    headers = [h.strip() for h in lines[0].split(",")]
    records = []
    for line in lines[1:]:
        vals = [v.strip() for v in line.split(",")]
        obj: dict[str, str] = {}
        for i, h in enumerate(headers):
            obj[h] = vals[i] if i < len(vals) else ""
        records.append(obj)
    return {"records": records, "columns": headers}


def _parse_jsonl_table(content: str) -> dict[str, Any]:
    lines = [l for l in content.strip().split("\n") if l.strip()]
    records = [json.loads(line) for line in lines]
    columns = list(records[0].keys()) if records else []
    return {"records": records, "columns": columns}


def _infer_column_types(
    records: list[dict[str, Any]], columns: list[str]
) -> list[dict[str, Any]]:
    if not records:
        return records

    col_types: dict[str, str] = {}
    for col in columns:
        col_type = "string"
        all_empty = True
        for rec in records:
            raw_val = rec.get(col, "")
            # Normalize to lowercase string (matches JS String() behavior)
            val = str(raw_val).strip() if not isinstance(raw_val, bool) else str(raw_val).lower()
            if val == "":
                continue
            all_empty = False
            if val in ("true", "false"):
                if col_type == "string":
                    col_type = "boolean"
            else:
                try:
                    num = float(val)
                    if col_type == "string":
                        col_type = "integer" if num == int(num) and "." not in val else "float"
                except ValueError:
                    col_type = "string"
                    break
        if all_empty:
            col_type = "string"
        col_types[col] = col_type

    result = []
    for rec in records:
        out: dict[str, Any] = {}
        for col in columns:
            raw_val = rec.get(col, "")
            # For booleans, use the native value directly
            if isinstance(raw_val, bool) and col_types[col] == "boolean":
                out[col] = raw_val
                continue
            raw = str(raw_val).strip() if not isinstance(raw_val, bool) else str(raw_val).lower()
            ct = col_types[col]
            if ct == "boolean":
                out[col] = raw == "true"
            elif ct == "integer":
                out[col] = int(raw) if raw else ""
            elif ct == "float":
                out[col] = float(raw) if raw else ""
            else:
                out[col] = raw
        result.append(out)
    return result


def _object_to_xml(data: Any, indent: str = "", wrap_in_item: bool = False) -> str:
    if isinstance(data, list):
        return "\n".join(_object_to_xml(item, indent, True) for item in data)
    if isinstance(data, dict):
        child_indent = indent + "  " if wrap_in_item else indent
        parts = []
        for k, val in data.items():
            if val is not None and isinstance(val, (dict, list)):
                inner = _object_to_xml(val, child_indent + "  ")
                parts.append(f"{child_indent}<{k}>\n{inner}\n{child_indent}</{k}>")
            else:
                str_val = _js_string(val)
                if str_val == "":
                    parts.append(f"{child_indent}<{k}/>")
                else:
                    parts.append(f"{child_indent}<{k}>{_escape_xml_content(str_val)}</{k}>")
        inner = "\n".join(parts)
        if wrap_in_item:
            return f"{indent}<item>\n{inner}\n{indent}</item>"
        return inner
    return indent + _escape_xml_content(_js_string(data))


def _escape_xml_content(s: str) -> str:
    return s.replace("&", "&amp;").replace("<", "&lt;").replace(">", "&gt;")


def _js_string(val: Any) -> str:
    """Convert a value to string matching JavaScript's String() behavior."""
    if val is None:
        return ""
    if isinstance(val, bool):
        return "true" if val else "false"
    return str(val)
